java/src/Ice/src/main/java/com/zeroc/Ice/ConnectionI.java

```diff
@@ -1220,17 +1222,80 @@ public com.zeroc.IceInternal.ThreadPool getThreadPool() {
     return _threadPool;
   }
 
+  public synchronized void idleCheck(
+      int idleTimeout, BooleanSupplier isTimerTaskStarted, Runnable rescheduleTimer) {
+    // If isTimerTaskStarted returns false, it means that while we were waiting to acquire the
+    // lock, a read went through and rescheduled the read timer task. This means "this" task was
+    // canceled so we don't do anything.
+    if ((_state == StateActive || _state == StateHolding) && isTimerTaskStarted.getAsBoolean()) {
+      if (_transceiver.isWaitingToBeRead()) {
+        // Bytes are available for reading but the thread pool is exhausted. We don't want to abort
+        // the connection in this situation.
+        rescheduleTimer.run();
+
+        if (_instance.traceLevels().network >= 3) {
+          _instance
+              .initializationData()
+              .logger
+              .trace(
+                  _instance.traceLevels().networkCat,
+                  "the idle check scheduled a new idle check in "
+                      + idleTimeout
+                      + "s because the connection is waiting to be read\n"
+                      + _transceiver.toDetailedString());
+        }
+      } else {
+        if (_instance.traceLevels().network >= 1) {
+          _instance
+              .initializationData()
+              .logger
+              .trace(
+                  _instance.traceLevels().networkCat,
+                  "connection aborted by the idle check because it did not receive any byte for "
+                      + idleTimeout
+                      + "s\n"
+                      + _transceiver.toDetailedString());
+        }
+
+        setState(
+            StateClosed,
+            new ConnectionTimeoutException()); // TODO: should be ConnectionIdleException
+      }
+    }
+    // else nothing to do
+  }
+
+  public synchronized void sendHeartbeat() {
+    assert !_endpoint.datagram();
+
+    if (_state == StateActive || _state == StateHolding) {
```

Comment:
We have a method for checking this, whicih we should consider using here:
```suggestion
    if (isActiveOrHolding()) {
```
Or maybe we should just delete the method... because it's kind of crazy to have

Reply:
Fixed.